core/lib/zksync_core/src/basic_witness_input_producer/mod.rs

```diff
@@ -0,0 +1,307 @@
+use std::sync::Arc;
+use std::time::Instant;
+
+use crate::state_keeper::io::common::load_l1_batch_params;
+
+use multivm::VmInstance;
+use vm::{HistoryEnabled, L2BlockEnv};
+use zksync_config::configs::chain::{NetworkConfig, StateKeeperConfig};
+use zksync_config::constants::{
+    SYSTEM_CONTEXT_ADDRESS, SYSTEM_CONTEXT_CURRENT_L2_BLOCK_INFO_POSITION,
+};
+use zksync_dal::{ConnectionPool, StorageProcessor};
+use zksync_object_store::{ObjectStore, ObjectStoreFactory};
+use zksync_queued_job_processor::JobProcessor;
+use zksync_state::{PostgresStorage, PostgresStorageCaches, ReadStorage, StorageView};
+use zksync_types::block::unpack_block_info;
+use zksync_types::witness_block_state::WitnessBlockState;
+use zksync_types::{
+    AccountTreeId, L1BatchNumber, L2ChainId, MiniblockNumber, StorageKey, Transaction,
+};
+use zksync_utils::h256_to_u256;
+
+use anyhow::Context;
+use async_trait::async_trait;
+use tokio::task::JoinHandle;
+
+#[derive(Debug)]
+pub struct BasicWitnessInputProducerJob {
+    l1_batch_number: L1BatchNumber,
+}
+
+async fn get_miniblock_transition_state(
+    connection: &mut StorageProcessor<'_>,
+    miniblock_number: MiniblockNumber,
+) -> L2BlockEnv {
+    let l2_block_info_key = StorageKey::new(
+        AccountTreeId::new(SYSTEM_CONTEXT_ADDRESS),
+        SYSTEM_CONTEXT_CURRENT_L2_BLOCK_INFO_POSITION,
+    );
+    let l2_block_info = connection
+        .storage_web3_dal()
+        .get_historical_value_unchecked(&l2_block_info_key, miniblock_number + 1)
+        .await
+        .unwrap();
+
+    let (next_miniblock_number, next_miniblock_timestamp) =
+        unpack_block_info(h256_to_u256(l2_block_info));
+
+    let miniblock_hash = connection
+        .blocks_web3_dal()
+        .get_miniblock_hash(miniblock_number)
+        .await
+        .unwrap()
+        .unwrap();
+
+    let virtual_blocks = connection
+        .blocks_dal()
+        .get_virtual_blocks_for_miniblock(&(miniblock_number + 1))
+        .await
+        .unwrap()
+        .unwrap();
+
+    L2BlockEnv {
+        number: next_miniblock_number as u32,
+        timestamp: next_miniblock_timestamp,
+        prev_block_hash: miniblock_hash,
+        max_virtual_blocks_to_create: virtual_blocks,
+    }
+}
+
+fn execute_tx<S: ReadStorage>(tx: &Transaction, vm: &mut VmInstance<S, HistoryEnabled>) {
+    vm.make_snapshot();
+    if vm
+        .inspect_transaction_with_bytecode_compression(vec![], tx.clone(), true)
+        .is_ok()
+    {
+        vm.pop_snapshot_no_rollback();
+        return;
+    }
```

Comment:
oh this is a little hard to follow (I can imagine it was confusing to write as well) - let's add some comments maybe?